src/intptrcast.rs

```diff
@@ -197,17 +197,20 @@ impl<'mir, 'tcx: 'mir> EvalContextExt<'mir, 'tcx> for crate::MiriInterpCx<'mir,
 pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
     fn expose_ptr(&mut self, alloc_id: AllocId, tag: BorTag) -> InterpResult<'tcx> {
         let ecx = self.eval_context_mut();
-        // Make sure we aren't trying to expose a dead allocation; this is part of justifying the
-        // way we treat the exposed set in VisitProvenance.
-        assert!(!matches!(ecx.get_alloc_info(alloc_id).2, AllocKind::Dead));
+        // Exposing a dead alloc is a no-op, because it's not possible to get a dead allocation
+        // via int2ptr.
+        if matches!(ecx.get_alloc_info(alloc_id).2, AllocKind::Dead) {
```

Comment:
Please check this *after* `provenance_mode`. I doubt it matters, but still -- the cheap check should come first, and `get_alloc_info` can potentially require a query.